### ISSUE TITLE
Refractor to improve the efficiency of markbind with parallelism

### ISSUE THREAD

**Issue by yiwen101** [MarkBind/markbind#2503]
Some of the current code can benefit from parallelism. Eg:

<img width="631" alt="Screenshot 2024-04-06 at 10 28 35" src="https://github.com/MarkBind/markbind/assets/121547057/62e25bd5-cd7e-448b-b238-58cd3805cefa">

Another issue of this way of writing is, we do not have a clear idea on the dependency of the sequence of run. 

In future,  we should run every no conflicting async function in parallel, and shift to A.then().then() style to specify sequential dependencies.

**Comment by tlylt** [comment]
PRs that address this issue should be careful with the potential unintended impacts on functionality. I believe some functions have (undesirable) implicit dependencies and introducing concurrent execution could lead to subtle bugs. The prereq for tackling this is therefore to come up with a clear flow diagram describing the various operations.

Since this is about performance, another thing to ensure is to capture the before and after timing, as well as CPU/memory consumption to assess the impact.
